### PR TITLE
Numericality validation skips some checks when a proc or a method call return nil

### PR DESCRIPTION
**Pull request by DmitryTsepelev**
Numericality validation should skip `:greater_than`, `:greater_than_or_equal_to`, `:equal_to`,
`:less_than`, `:less_than_or_equal_to` checks when a proc or a method call return nil instead
of throwing `ArgumentError: comparison of Integer with nil failed`.

In my opinion, it would be more helpful to skip these checks, because in this case user won't have to write a complex proc to handle the case when the value it's compared to is nil.

In other words, now we can write:

```ruby
class ProductFilter
  include ActiveModel::Validations

  validates :min_price, numericality: { allow_nil: true }
  validates :max_price, numericality: { allow_nil: true, greater_than: :min_price }
end

ProductFilter.new(max_price: 100).valid? # => true
```

instead of something like:

```ruby
class ProductFilter
  include ActiveModel::Validations

  validates :min_price, numericality: { allow_nil: true }
  validates :max_price, numericality: { allow_nil: true, greater_than: ->(filter) { filter.min_price.to_i } }
end

ProductFilter.new(max_price: 100).valid? # => true
```

### CODE REVIEW

**Review by bogdanvlviv — APPROVED** [review]

File: activemodel/lib/active_model/validations/numericality.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@ class NumericalityValidator < EachValidator # :nodoc:
       CHECKS = { greater_than: :>, greater_than_or_equal_to: :>=,
                  equal_to: :==, less_than: :<, less_than_or_equal_to: :<=,
                  odd: :odd?, even: :even?, other_than: :!= }.freeze
-
+      CHECKS_IGNORING_NIL = %i(greater_than greater_than_or_equal_to equal_to less_than less_than_or_equal_to)
```
Comment (bogdanvlviv) [inline activemodel/lib/active_model/validations/numericality.rb]: I think we can remove `equal_to` from the ignore list

Comment (DmitryTsepelev) [reply activemodel/lib/active_model/validations/numericality.rb]: Well, I think you're right and I should remove it - since it does not compare numbers it won't crash in a way described above. Also, it would cause a following confusing behavior:

```ruby
class Smth
  include ActiveModel::Validations
  validates :field_one, numericality: { allow_nil: true }
  validates :field_two, numericality: { allow_nil: true, equal_to: :field_one }
end

Smth.new(field_two: 1).valid? # => true, but why? these two are not equal
```

Removing that away, thanks for the review @bogdanvlviv

File: activemodel/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,22 @@
+*   Numericality validation should skip `:greater_than`, `:greater_than_or_equal_to`, `:equal_to`,
```
Comment (bogdanvlviv) [inline activemodel/CHANGELOG.md]: We should remove `:equal_to` from here as well.

> Well, I think you're right and I should remove it - since it does not compare numbers it won't crash in a way > described above. Also, it would cause a following confusing behavior:
> class Smth
>   include ActiveModel::Validations
>   validates :field_one, numericality: { allow_nil: true }
>   validates :field_two, numericality: { allow_nil: true, equal_to: :field_one }
> end
> 
> Smth.new(field_two: 1).valid? # => true, but why? these two are not equal

We should probably add test case for that in order to prevent any regression in the feature and describe the existing behavior

File: activemodel/lib/active_model/validations/numericality.rb [hunk 3]
```diff
@@ -184,6 +186,9 @@ module HelperMethods
       #     validates_numericality_of :width, less_than: ->(person) { person.height }
       #     validates_numericality_of :width, greater_than: :minimum_weight
       #   end
+      #
+      # A check passes when a proc or a method call return nil.
```
Comment (bogdanvlviv) [inline activemodel/lib/active_model/validations/numericality.rb]: We should mention that a check passes unless `:equal_to`

File: activemodel/CHANGELOG.md (hunk 2 above)
Comment (DmitryTsepelev) [reply activemodel/CHANGELOG.md]: Agreed, added one more test case

File: activemodel/lib/active_model/validations/numericality.rb (hunk 3 above)
Comment (DmitryTsepelev) [reply activemodel/lib/active_model/validations/numericality.rb]: ✔️

**Comment by bogdanvlviv** [comment]
LGTM! Thanks @DmitryTsepelev.

**Comment by DmitryTsepelev** [comment]
@kamipo maybe you could take a look at it?

**Comment by DmitryTsepelev** [comment]
@rafaelfranca Sorry for pinging you again, just wanted to check if we could get this PR reviewed since RC1 is already in the separate branch

**Comment by bogdanvlviv** [comment]
@DmitryTsepelev Could you rebase to resolve conflicts in `activemodel/CHANGELOG.md` and see whether full test suite run is still being succeeded after rebasing.

**Comment by DmitryTsepelev** [comment]
@bogdanvlviv Looks like it's still passing 🙂

File: activemodel/test/cases/validations/numericality_validation_test.rb [hunk 4]
```diff
@@ -233,6 +233,31 @@ def test_validates_numericality_with_symbol
     Topic.remove_method :max_approved
   end
 
+  def test_validates_numericality_with_symbol_and_nil_value
+    Topic.define_method(:max_approved) { nil }
+    Topic.validates_numericality_of :approved, less_than_or_equal_to: :max_approved
+
+    valid!([4])
```
Comment (kamipo) [inline activemodel/test/cases/validations/numericality_validation_test.rb]: Looks this spec (`4 <= nil` is valid) is too application specific to me.

Comment (DmitryTsepelev) [reply activemodel/test/cases/validations/numericality_validation_test.rb]: That's probably true – I cannot come up with a different example rather than a pair of `min` and `max` fields, but `ArgumentError` seemed confusing to me and a workaround to fix that (`greater_than: ->(filter) { filter.min_price.to_i }`) looks a bit wordy

Comment (DmitryTsepelev) [reply activemodel/test/cases/validations/numericality_validation_test.rb]: @kamipo how can I make this spec look better?

**Comment by coorasse** [comment]
related to https://github.com/rails/rails/issues/27968, from the discussion I see there, @rafaelfranca didn't approve this approach (and I agree 😄  ).

**Comment by DmitryTsepelev** [comment]
@coorasse thanks for bringing it up! I think my example is a bit different: there is comparison to the value, where nil is _allowed_, and I don't mind the exception to be risen when nil is returned by mistake 🙂 Sometimes it makes sense to run the validation only when the value to compare with is present (e.g., in my example with price range)

**Comment by coorasse** [comment]
I know the struggle because I  stumbled upon exactly this issue already three or four times, and every single time I had to change from
`validates :max_price, numericality: { allow_nil: true, greater_than: :min_price }`

to

`validates :max_price, numericality: { allow_nil: true, greater_than: ->(filter) { filter.min_price.to_i } }`

and it's quite annoying, I agree 100%. 😆 

On the other side, I also think that these are two different validations, and in the second one you are managing the `nil` case explicitly, relying on the fact that `nil.to_i` returns 0 and therefore your validation will pass. 
If we would be playing also with negative numbers, you'd have a problem again: so this is not a 100% proof solution, still.

To me, we should have a `rubocop-rails` check, where such a case is identified and tells you "you are comparing with something which might be nil. this might cause an error at runtime, please use a proc instead".
